File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.py
from collections import deque

def bfs(start, graph, link_check, visited):
    que = deque([start])
    visited[start] = True
    cnt = 1
    while que:
        s = que.popleft()
        for v in graph[s]:
            if visited[v] == False and link_check[start][v]:
                que.append(v)
                visited[v] = True
                cnt += 1
    return cnt

def solution(n, wires):
    answer = n
    link_check = [[True]*(n+1) for _ in range(n+1)]
    graph = [[] for _ in range(n+1)]
    
    for a, b in wires:
        graph[a].append(b)
        graph[b].append(a)
        
    for a, b in wires:
        visited = [False]*(n+1)
        link_check[a][b] = False
        cnt_a = bfs(a, graph, link_check, visited)
        cnt_b = bfs(b, graph, link_check, visited)
        link_check[a][b] = True
        
        answer = min(answer, abs(cnt_a-cnt_b))
        
    return answer

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.java
import java.util.*;

class Solution {
    static ArrayList<ArrayList<Boolean>> link = new ArrayList<>();
    static ArrayList<ArrayList<Integer>> graph = new ArrayList<>();
    static boolean[] visit;
    public int solution(int n, int[][] wires) {
        int answer = n;
        for (int i = 0; i <= n; i++) {
            link.add(new ArrayList<>(Collections.nCopies(n+1, true)));
            graph.add(new ArrayList<>());
        }
        
        for (int[] w : wires) {
            graph.get(w[0]).add(w[1]);
            graph.get(w[1]).add(w[0]);
        }
        for (int[] w : wires) {
            visit = new boolean[n+1];
            link.get(w[0]).set(w[1], false);
            link.get(w[1]).set(w[0], false);
            int cnt_a = bfs(w[0]);
            int cnt_b = bfs(w[1]);
            answer = Math.min(answer, Math.abs(cnt_a - cnt_b));
            link.get(w[0]).set(w[1], true);
            link.get(w[1]).set(w[0], true);
        }
        return answer;
    }
    
    static int bfs(int start) {
        Queue<Integer> que = new ArrayDeque<>();
        que.add(start);
        visit[start] = true;
        int cnt = 1;
        
        while (!que.isEmpty()) {
            int cur = que.poll();
            for (int a : graph.get(cur)) {
                if (!visit[a] && link.get(cur).get(a)) {
                    visit[a] = true;
                    que.add(a);
                    cnt++;
                }
            }
        }
        return cnt;
    }
}